Mathlib/Order/Filter/Basic.lean

```diff
@@ -0,0 +1,3090 @@
+which is a special case of `mem_closure_of_tendsto` from topology.basic.
+
+## Notations
+
+* `∀ᶠ x in f, p x` : `f.Eventually p`;
+* `∃ᶠ x in f, p x` : `f.Frequently p`;
+* `f =ᶠ[l] g` : `∀ᶠ x in l, f x = g x`;
+* `f ≤ᶠ[l] g` : `∀ᶠ x in l, f x ≤ g x`;
+* `𝓟 s` : `Filter.Principal s`, localized in `Filter`.
+
+## References
+
+*  [N. Bourbaki, *General Topology*][bourbaki1966]
+
+Important note: Bourbaki requires that a filter on `X` cannot contain all sets of `X`, which
+we do *not* require. This gives `filter X` better formal properties, in particular a bottom element
+`⊥` for its lattice structure, at the cost of including the assumption
+`[NeBot f]` in a number of lemmas and definitions.
+-/
+
+
+open Function Set Order
+open Classical hiding by_cases not_not
+
+universe u v w x y
+
+/-- A filter `F` on a type `α` is a collection of sets of `α` which contains the whole `α`,
+is upwards-closed, and is stable under intersection. We do not forbid this collection to be
+all sets of `α`. -/
+structure Filter (α : Type _) where
+  /-- The set of sets that belong to the filter. -/
+  sets : Set (Set α)
+  /-- The set `Set.univ` belongs to any filter. -/
+  univ_sets : Set.univ ∈ sets
+  /-- If a set belongs to a filter, then its superset belongs to the filter as well. -/
+  sets_of_superset {x y} : x ∈ sets → x ⊆ y → y ∈ sets
+  /-- If two sets belong to a filter, then their intersection belongs to the filter as well. -/
+  inter_sets {x y} : x ∈ sets → y ∈ sets → x ∩ y ∈ sets
+#align filter Filter
+
+/-- If `F` is a filter on `α`, and `U` a subset of `α` then we can write `U ∈ F` as on paper. -/
+instance {α : Type _} : Membership (Set α) (Filter α) :=
+  ⟨fun U F => U ∈ F.sets⟩
+
+namespace Filter
+
+variable {α : Type u} {f g : Filter α} {s t : Set α}
+
+@[simp]
+protected theorem mem_mk {t : Set (Set α)} {h₁ h₂ h₃} : s ∈ mk t h₁ h₂ h₃ ↔ s ∈ t :=
+  Iff.rfl
+#align filter.mem_mk Filter.mem_mk
+
+@[simp]
+protected theorem mem_sets : s ∈ f.sets ↔ s ∈ f :=
+  Iff.rfl
+#align filter.mem_sets Filter.mem_sets
+
+instance inhabitedMem : Inhabited { s : Set α // s ∈ f } :=
+  ⟨⟨univ, f.univ_sets⟩⟩
+#align filter.inhabited_mem Filter.inhabitedMem
+
+theorem filter_eq : ∀ {f g : Filter α}, f.sets = g.sets → f = g
+  | ⟨_, _, _, _⟩, ⟨_, _, _, _⟩, rfl => rfl
+#align filter.filter_eq Filter.filter_eq
+
+theorem filter_eq_iff : f = g ↔ f.sets = g.sets :=
+  ⟨congr_arg _, filter_eq⟩
+#align filter.filter_eq_iff Filter.filter_eq_iff
+
+protected theorem ext_iff : f = g ↔ ∀ s, s ∈ f ↔ s ∈ g := by
+  simp only [filter_eq_iff, ext_iff, Filter.mem_sets]
+#align filter.ext_iff Filter.ext_iff
+
+@[ext]
+protected theorem ext : (∀ s, s ∈ f ↔ s ∈ g) → f = g :=
+  Filter.ext_iff.2
+#align filter.ext Filter.ext
+
+/-- An extensionality lemma that is useful for filters with good lemmas about `sᶜ ∈ f` (e.g.,
+`filter.comap`, `filter.coprod`, `filter.Coprod`, `filter.cofinite`). -/
```

Comment:
```suggestion
`Filter.comap`, `Filter.coprod`, `Filter.Coprod`, `Filter.cofinite`). -/
```
Should this be called `Filter.Coprod`, or rather something like `Filter.coprodᵢ`?